crates/plugins/src/manager.rs

```diff
@@ -0,0 +1,219 @@
+    /// not have a conflicting name with Spin internal commands, and is not a downgrade of a
+    /// currently installed plugin.
+    pub fn check_manifest(
+        &self,
+        plugin_manifest: &PluginManifest,
+        spin_version: &str,
+        allow_downgrades: bool,
+    ) -> Result<()> {
+        // Disallow installing plugins with the same name as spin internal subcommands
+        if SPIN_INTERNAL_COMMANDS
+            .iter()
+            .any(|&s| s == plugin_manifest.name())
+        {
+            bail!(
+                "Can't install a plugin with the same name ({}) as an internal command",
+                plugin_manifest.name()
+            );
+        }
+
+        // Disallow reinstalling identical plugins and downgrading unless permitted.
+        if let Ok(installed) = self.store.read_plugin_manifest(&plugin_manifest.name()) {
+            if &installed == plugin_manifest {
+                bail!(
+                    "Plugin {} is already installed with version {}.",
+                    plugin_manifest.name(),
+                    installed.version,
+                );
+            } else if installed.version > plugin_manifest.version && !allow_downgrades {
+                bail!(
+                    "Newer version {} of {} plugin is already installed. To downgrade to version {} set the `--downgrade` flag.",
+                    installed.version,
+                    plugin_manifest.name(),
+                    plugin_manifest.version,
+                );
+            }
+        }
+
+        check_supported_version(plugin_manifest, spin_version)
+    }
+
+    /// Fetches a manifest from a local, remote, or repository location and returned the parsed
+    /// PluginManifest object.
+    pub async fn get_manifest(
+        &self,
+        manifest_location: &ManifestLocation,
+    ) -> Result<PluginManifest> {
+        let plugin_manifest = match manifest_location {
+            ManifestLocation::Remote(url) => {
+                log::info!("Pulling manifest for plugin from {}", url);
+                reqwest::get(url.as_ref())
+                    .await?
+                    .error_for_status()
+                    .with_context(|| format!("Manifest endpoint {} is unhealthy", url))?
+                    .json::<PluginManifest>()
+                    .await
+                    .with_context(|| format!("Failed to deserialize remote plugin at {}", url))?
+            }
+            ManifestLocation::Local(path) => {
+                log::info!("Pulling manifest for plugin from {}", path.display());
+                let file = File::open(path).with_context(|| {
+                    format!("The local manifest {} could not be opened", path.display())
+                })?;
+                serde_json::from_reader(file).with_context(|| format!("Failed to deserialize local manifest at {}. Ensure it has the expected plugin manifest format", path.display()))?
```

Comment:
Unless I'm missing something, `.display()` should be redundant here, implied by `{}`. Also you can (optionally) inline simple variables if you like:

```suggestion
                log::info!("Pulling manifest for plugin from {path}");
                let file = File::open(path).with_context(|| {
                    format!("The local manifest {path} could not be opened")
                })?;
                serde_json::from_reader(file).with_context(|| format!("Failed to deserialize local manifest at {path}. Ensure it has the expected plugin manifest format"))?
```

Reply:
Interesting point! I looked into this because casting to `Path` display type does not work and gets an error that path does not implement Display. Rather it still has a `display()` method that can be used that works around potential Unicode errors. `"{path}"` is disallowed and `path.display()` is the method encouraged by the docs: https://doc.rust-lang.org/stable/std/path/struct.Display.html

Reply:
>  Unless I'm missing something

I was missing something!

Reply:
Looking at some of my own code I realized why I was confused; I almost always use Debug output for logging/errors, e.g. `{path:?}` which _is_ supported by `Path`. The Debug formatter adds quotes and escapes special chars which may or not be desirable in different situations.